### ISSUE TITLE
After "Save As", projects believe they are unsaved

### ISSUE THREAD

**Issue by cdsmith** [google/codeworld#1065]
After a Save As of a new project, the newly saved project shows up in the dirty state.  Furthermore, clicking Save again pops up a new Save As dialog instead of saving with the current name.  Choosing to overwrite seems to create a second project in the list by the same name.

**Comment by misterbeebee** [comment]
Some more notes:
If you Save a 

The root of the issue may be that the New (and Dirty) state is carried over to a newly created (Save As) project. If you immediately save your initial empty project, you get a project that is not dirty, but Save still acts like Save As

You can see another symptom of this in Delete:  If you open a fresh workspace, and immediately Save (As), then the Delete button (on your newly Saved project) is a no-op.

This gives rise to a workaround:

1. Save (As) your first project as "empty". (Recommended: Do this immediately when launching CodeWorld)
2. Save (As) your (wrongly New) "empty" project to "project"
3. Switch to "empty" (OK to discard your non-existent changes.)  By this point CodeWorld no longer treats "empty" as New.
4. Optional: Delete "empty".
5. Switch back to "project".  By this point CodeWorld no longer treats "project" as New.
6. Everything works fine now... until you click "New", and your New project has the original bug.

Potential fix, which would also make the UI more clear:
  "New" (including the automatic New that comes with a new workspace) should create a project and immediately save it as "Untitled" or "New Project" or (Force the "Save As" UI), (or if you want to get fancy, "Untitled <N>" to avoid conflicts with other untitled.

**Comment by cdsmith** [comment]
Update: I looked into this last weekend  The root of the problem is that Save As just never updates `window.openProjectName` or `window.savedGeneration` at all.  However, it does add a new project to the tree view and display it as if it were selected.  This out-of-sync state between the tree state and the global variables is responsible for this bug, as well as #1066.  I wish I hadn't gotten into this mess; I was too eager to merge a pull request to encourage a new contributor.

I'll tackle this over the weekend.  Fixing this one case of loss-of-sync shouldn't be too hard.  But at the same time, I should also think about how to more systematically keep them in sync, perhaps by diffing and changing the tree control's model in the same places that we used to rebuild the project buttons.
